### PR TITLE
Try to set the Project's repo if it does not have one when a new Reposit...

### PR DESCRIPTION
**Pull request by bolinfest**
...oryProvider is registered.

I tested this using my test `HgRepositoryProvider`. Now when I run the following from the
command line:

```
atom <path-to-directory-with-hg-repository>
```

And then run the following in the console:

```
atom.project.getRepositories()
```

I get an array with an `HgRepository` in it. Previously, I got an empty array because the
`Project`'s paths were set before my `HgRepositoryProvider` was registered.

### CODE REVIEW

**Comment by maxbrunsfeld** [comment]
@bolinfest this seems pretty important. Can you add a test covering this case?

**Comment by bolinfest** [comment]
Absolutely. Is there an easy way to create an atom package with the test
Project that has the metadata that would trigger this? Or should I just
make this logic exercisable from a "visible for testing" method?

**Comment by maxbrunsfeld** [comment]
For now, could you just test it by using the private `atom.packages.serviceHub` variable to provide a dummy repo-provider?

**Comment by bolinfest** [comment]
Blargh, `spec-helper.coffee` does:

```
  atom.project = new Project(paths: [projectPath])
  atom.workspace = new Workspace()
  atom.packages.serviceHub = new ServiceHub
```

so it resets `atom.packages.serviceHub` after `atom.project` has just used it. I'll see if reordering these lines breaks anything.

**Comment by bolinfest** [comment]
Fortunately all tests still pass with:

```
  atom.packages.serviceHub = new ServiceHub
  atom.project = new Project(paths: [projectPath])
  atom.workspace = new Workspace()
```

**Comment by bolinfest** [comment]
OK, added test cases both for when there are paths without repositories and when there aren't.

File: spec/project-spec.coffee [hunk 1]
```diff
@@ -13,6 +13,37 @@ describe "Project", ->
...
+      repositoryProvider =
+        repositoryForDirectory: (directory) -> Promise.resolve(dummyRepository)
+        repositoryForDirectorySync: (directory) -> dummyRepository
+      atom.packages.serviceHub.provide(
+        "atom.repository-provider", "0.1.0", repositoryProvider)
+
+      expect(atom.project.repositoryProviders.length).toBe 2
+      expect(atom.project.getRepositories()).toEqual [dummyRepository]
+
+    it "does not update @repositories if every path has a Repository", ->
```
Comment (maxbrunsfeld) [inline spec/project-spec.coffee]: Is it true that every path has a repository in this case? It looks like the path is being set to `spec/fixtures/dir` here, which doesn't contain a git repo.

Comment (bolinfest) [inline spec/project-spec.coffee]: Yep, exactly. /tmp does not have a repository initially and then the
provider supplies one.

Comment (bolinfest) [inline spec/project-spec.coffee]: Oh, in the second case, I thought that atom.project points to the Atom repo
itself by default, but I should confirm.

Comment (maxbrunsfeld) [inline spec/project-spec.coffee]: It points to the `spec/fixtures` directory by default, but it is being set to `spec/fixtures/dir` in [the beforeEach](https://github.com/atom/atom/pull/5537/files#diff-f42f793b4715cfbf8d07c772bc142849L14) of the test.

**Comment by bolinfest** [comment]
@maxbrunsfeld You're right. But `GitRepositoryProvider` takes a `Directory` and searches upwards until it finds a parent with a `.git` subdirectory or the root of the file tree. So in this case, it finds the Atom repo.

**Comment by maxbrunsfeld** [comment]
Got it :+1:. Thanks for adding that extra assertion in there in case anyone else is confused about that.

**Comment by bolinfest** [comment]
Yep, of course. Let me know if there are any other scenarios that you think should be tested.

**Comment by bolinfest** [comment]
Also, if we could get this in before the 0.180.0 release, I'd be much
obliged.

File: spec/project-spec.coffee [hunk 2]
```diff
@@ -13,6 +13,39 @@ describe "Project", ->
...
+      repositories = atom.project.getRepositories()
+      expect(repositories.length).toEqual 1
+      [repository] = repositories
+      expect(repository).toBeTruthy()
+
+      # Register a new RepositoryProvider.
+      dummyRepository = destroy: () ->
```
Comment (kevinsawicki) [inline spec/project-spec.coffee]: :fire: `()`

File: spec/project-spec.coffee [hunk 3]
```diff
@@ -13,6 +13,39 @@ describe "Project", ->
   beforeEach ->
     atom.project.setPaths([atom.project.getDirectories()[0]?.resolve('dir')])
 
+  describe "constructor", ->
+    it "tries to update repositories when a new RepositoryProvider is registered", ->
+      atom.project.setPaths(["/tmp"])
```
Comment (kevinsawicki) [inline spec/project-spec.coffee]: This shouldn't use a hard-coded platform-specific path, it should use something like `temp.mkdirSync('atom-project')` instead.

File: spec/project-spec.coffee [hunk 4]
```diff
@@ -13,6 +13,39 @@ describe "Project", ->
...
+      expect(atom.project.getRepositories()).toEqual [null]
+      expect(atom.project.repositoryProviders.length).toEqual 1
+
+      # Register a new RepositoryProvider.
+      dummyRepository = destroy: () ->
```
Comment (kevinsawicki) [inline spec/project-spec.coffee]: :fire: `()`

File: spec/project-spec.coffee (hunk 2 above)
Comment (bolinfest) [inline spec/project-spec.coffee]: done

File: spec/project-spec.coffee (hunk 3 above)
Comment (bolinfest) [inline spec/project-spec.coffee]: done

File: spec/project-spec.coffee (hunk 4 above)
Comment (bolinfest) [inline spec/project-spec.coffee]: done

**Pull request merged**